enhancements/management-workload-partitioning.md

```diff
@@ -0,0 +1,666 @@
+to the radio antenna for which it is responsible. In practice, this
+can mean running this component on anything from a single server at
+the base of a remote cell tower or in a datacenter-like environment
+serving several base stations.
+
+A hypothetical DU example is an unusually resource-intensive workload,
+requiring 20 dedicated cores, 24 GiB of RAM consumed as huge pages,
+multiple SR-IOV NICs carrying several Gbps of traffic each, and
+specialized accelerator devices. The node hosting this workload must
+run a realtime kernel, be carefully tuned to ensure low-latency
+requirements can be met, and be configured to support features like
+Precision Timing Protocol (PTP).
+
+The most constrained resource for RAN deployments is CPU, so for now
+the focus of this enhancement is managing CPU requirements.
+Kubernetes resource requests for CPU time are measured in fractional
+"[Kubernetes
+CPUs](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#meaning-of-cpu)". For
+bare metal, 1 Kubernetes CPU is equivalent to 1 hyperthread.
+
+Due to the resource-intensive workload, the overhead for platform
+components such as the OpenShift platform and add-on operators is
+severely restricted when compared to other uses. For example, one
+customer has allocated 4 hyperthreads for all components other than
+their own workloads. OpenShift currently requires 7.
+
+### Implementation Details/Notes/Constraints
+
+#### High-level End-to-end Workflow
+
+This section outlines an end-to-end workflow for deploying a cluster
+with management-workload-partitioning enabled and how pods are
+correctly scheduled to run on the management CPU pool.
+
+1. User sits down at their computer.
+2. The user creates their `install-config.yaml`, including extra
+   values in the specifying the CPUs to include in the management CPU
+   set. (See below for details.)
+   * Does the user also need to explicitly enable the feature flag?
+3. The user runs the installer.
+4. The installer uses the management CPU pool settings to generate an
+   extra machine config manifest to configure CRI-O to process
+   management workloads in a special way.
+5. The installer creates a machine config manifest to write a
+   configuration file for kubelet. The file should only be readable by
+   the kubelet.
+6. The installer enables the feature set for the feature flag
+   controlling management workload partitioning.
+7. The kubelet starts up and finds the configuration file enabling the
+   new feature.
+8. The kubelet advertises `cpu.workload.openshift.io/management`
+   extended resources on the node based on the number of CPUs in the
+   host.
+9. The kubelet reads static pod definitions. It replaces the `cpus`
+   requests with `cpu.workload.openshift.io/management` requests of
+   the same value and adds the
+   `cpu.workload.openshift.io/{container-name}` annotations for CRI-O
+   with the same values.
+10. Something schedules a regular pod with the
```

Comment:
I think we're likely to want to allow some pods in a namespace to target CPUs group/X and some pods in the same namespace to target CPUs group/Y.  In order to do this and maintain control, the namespace level label is like something like `workload.openhift.io/target-cpu-sets-allowed=management,in,comma,separate,list`.  And the pod level label is something like `workload.openshift.io/target-core-set=management`.

That is the solution I described in the e2e test that catches namespaces missing their labels: https://github.com/openshift/origin/pull/25992/files

Reply:
Is it better/more consistent to use 1 label with multiple values than to use multiple labels? I don't have a good sense of what's "usual" for that. Nodes can have multiple role labels, for example, and this feels similar to that.

Reply:
We discussed this a bit more in slack and I think I understand the reasoning for doing it this way. I'm not sure if we want the spelling exactly as proposed. We talk about CPU sets in this doc, but including that in the label implies that's the only aspect of workload isolation that we might ever do (I could see networking being used, for example).

How about `workload.openshift.io/allowed=comma,separated,list` for the permission label on namespaces and `workload.openshift.io/target=single-name` for the label on pods indicating the intent?

Reply:
I think the latest draft incorporates this comment thread. See 340d2dd62c1e669d68a59c76c8cc989386514340